### PR TITLE
i#5724 chunk bounds: Insert encodings in new chunks

### LINKED ISSUE #5724: invariant_checker failure on a64: "Chunk instruction counts are inconsistent"
This happened a couple of times in both debug and release builds:

http://139.178.82.61:8080/job/DynamoRIO-AArch64-Precommit/2313/consoleFull

```
262:   *** postcmd failed (Child aborted): Trace invariant failure in T418399 at
262:   ref # 93263: Chunk instruction counts are inconsistent
262: 
262:   ***
262: 
262: Call Stack (most recent call first):
262:   /var/lib/jenkins/workspace/DynamoRIO-AArch64-Precommit/suite/tests/runmulti.cmake:120 (process_cmdline)
262: 
262: 
285/317 Test #262: code_api|tool.drcacheoff.invariant_checker ......................***Failed    1.08
```

Some later instances.  They are always at ref 93263 it seems:

```
 Trace invariant failure in T498512 at
262:   ref # 93263: Chunk instruction counts are inconsistent
```
```
Trace invariant failure in T548743 at
259:   ref # 93263: Chunk instruction counts are inconsistent
```

### PR DESCRIPTION
**Pull request by derekbruening**
Adds logic to insert new encodings for instruction entries that were generated before it was known they'd appear after a chunk boundary. This has some inherent complexity due to the different modules owning record generation and chunk creation, along with the complexity of needing the decode PC for instruction records generated in the past.

Adds a unit test.

This finishes #5724, allowing us to remove it from the flaky list.

Fixes #5724

### CODE REVIEW

**Comment by derekbruening** [comment]
x64 failure is windows-zlib known issue #5635

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 1]
```diff
@@ -1249,38 +1255,77 @@ raw2trace_t::open_new_chunk(raw2trace_thread_data_t *tdata)
...
-                                              tdata->cur_chunk_instr_count);
-                    tdata->cpu2sched[it->addr].emplace_back(
-                        tdata->tid, tdata->last_timestamp_, tdata->last_cpu_,
-                        tdata->cur_chunk_instr_count);
-                }
-                continue;
-            }
             // We wait until we're past the final instr to write, to ensure we
             // get all its memrefs, by not stopping until we hit an instr or an
             // encoding.  (We will put function markers for entry in the
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/raw2trace.cpp:1324]: Unrelated to this PR: so we duplicate function markers? I remember a discussion on that recently. Is this the cause?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 2]
```diff
@@ -1291,15 +1336,61 @@ raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *e
...
-                // TODO i#5724: We need to re-emit encodings for "it" and any further
-                // instrs in this buffer: have a callback passed in which constructs
-                // an encoding from an instr record?
             }
-            if (type_is_instr(static_cast<trace_type_t>(it->type)))
+            if (type_is_instr(static_cast<trace_type_t>(it->type))) {
                 ++tdata->cur_chunk_instr_count;
+                ++instr_ordinal;
+                if (instr_ordinal >= static_cast<int>(decode_pcs.size()))
+                    return "decode_pcs is missing entries for written instructions";
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/raw2trace.cpp:1344]: This error-return should probably also be gated by `TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, tdata->file_type)`.

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 3]
```diff
@@ -1291,15 +1336,61 @@ raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *e
...
+            // identify missing post-chunk encodings?  Or should we have the reader
+            // deliberately clear its encoding history on a chunk boundary, raising
+            // a fatal error on a missing encoding?  For now the only complex case
+            // is these already-generated records which we handle here and have a
+            // unit test covering so those further checks are lower priority.
+            if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, tdata->file_type) &&
+                type_is_instr(static_cast<trace_type_t>(it->type)) &&
+                // We don't want encodings for the PC-only i-filtered entries.
+                it->size > 0 && !prev_was_encoding &&
+                record_encoding_emitted(tls, decode_pcs[instr_ordinal])) {
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/raw2trace.cpp:1358]: So this logic seems to properly handle branches that were delayed multiple times. Maybe comment about that?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 4]
```diff
@@ -1291,15 +1336,61 @@ raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *e
...
+                if (it > start &&
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/raw2trace.cpp:1359]: Also, for clarity, comment that if we see an instr that requires a delayed encoding, we write data from `start` now, or else it all gets written later after the loop ends.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp:1324]: No, we don't duplicate function markers.  Not sure which discussion: if you remember more we can discuss ofline.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp:1344]: Done, plus the size>0 check.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp:1358]: Done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.cpp:1359]: Done.

**Pull request merged**